git-lfs support

b-camacho opened the pull request:
Hi! This PR works, but there's definitely a better way to support LFS. Looking for maintainer feedback.

# Motivation
`nix` fetches git repos using `libgit2`, which does not run filters by default. This means LFS-enabled repos can be fetched, but LFS pointer files are not smudged. 

This change adds a `lfs` attribute to fetcher URLs. With `lfs=1`, when fetching LFS-enabled repos, nix will smudge all the files.

# Context
See https://github.com/NixOS/nix/issues/10079.
Git Large File Storage lets you track large files directly in git, using [git filters](https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes). A `clean` filter runs on your LFS-enrolled files before push, replacing large files with small "pointer files". Upon checkout, a "smudge" filter replaces pointer files with full file contents. When this works correctly, it is not visible to users, which is nice.

# Limitations
Right now, for repos with `lfs=1`, the LFS-tracked files are materialized during `nix flake lock` - this is bad, I'm looking for a way to avoid this.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

b-camacho commented:
Small complication:
it seems that `nix flake lock` calls [fetch](https://github.com/NixOS/nix/blob/8fb36a98ffae53683cb55165578adab21d8bbd8d/src/libfetchers/fetchers.hh#L79-L83) which in turn calls `Input::fetch` -> `InputScheme::fetch` -> `fetchToStore`.

In other words, I don't currently see a way to:
- materialize LFS files when fetching `-source` store paths, but
- don't materialize LFS files during `nix flake lock`.

L-as commented:
Why does Nix needs to support LFS? Isn't LFS typically for large  files,that wouldn't usually affect evaluation anyway?

b-camacho commented:
> What use case do you have in mind? Isn't LFS typically for large files, that wouldn't usually affect evaluation anyway?

`builtins.fetchGit` populates the nix store with a `<hash>-store` path. This path is used as the source when building a derivation. Currently, the builder will see the unsmudged LFS pointer files, but I'd like the builder to optionally see smudged files. I agree that smudged files are usually not needed at eval time, but I don't see a good alternative of making them available at build time, besides a fixed-output derivation (but fixed-output derivations have their own problems)

L-as commented:
A FOD seems optimal here, in general you shouldn't use builtins.fetchGit if you're only going to use it at build time.

b-camacho commented:
In general I agree, but (afaik) other fetchers can't use git credentials.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-03-11-nix-team-meeting-132/42960/1

khoitd1997 commented:
@roberth have you had a chance to take a look at this issue? We have been staying at older versions of Nix as a workaround but newer versions now have fixes for critical issues so sticking to old ones would no longer be optimal.

roberth commented:
Hi @b-camacho, thanks for the ping and sorry for the delay. This PR was assigned to me, but I hadn't prioritized it because it was a draft. Wrong assumption on my end, because I do think this is valuable, and I have some things to say :)

> when fetching LFS-enabled repos, nix will smudge all the files.

That's a good start, but we need to make sure that the smudging happens in a controlled manner; otherwise we risk adding impurities.

Specifically, we should parse the attribute to check that they're supposed to be unsmudged by lfs; if not, ignore the smudge rule. It seems you were already investigating how this could be implemented.

Furthermore, we should validate the sha256 so that we don't increase the potential for silent errors by a whole external program. The hash should be easy to parse from the pointer file, and while reading other programs' inputs is a little ad hoc, I don't expect any serious issues from this, as we won't cause users to accidentally rely on a bug this way.

--------

> the LFS-tracked files are materialized during `nix flake lock` - this is bad

This won't happen unnecessarily either of these are implemented
- #6530
- https://github.com/NixOS/nix/pull/10252

If we need to backtrack on the removal of narHashes (#6530), we can also avoid re-locking transitive inputs whose lock has already been computed by the dependency's lock.
- https://github.com/NixOS/nix/issues/7730

So yes, this isn't efficient yet, but it will be.




> A FOD seems optimal here, in general you shouldn't use `builtins.fetchGit` if you're only going to use it at build time.

A fixed output derivation works best when all you're using it for is as an input to another derivation (*and* it's publicly available, as mentioned).
However, if the "fetched" source is a flake (e.g. you have a `flake.nix` in a repo with LFS files), then you also need to evaluate files from the fetched source, which would constitute import from derivation, which is not optimal. Furthermore you'd need to produce fixed-output hashes for your local repo files, which is such horrible UX we don't need to consider it as a solution.

-----------

To summarize, this is worth implementing, I see no blocking issues, design or otherwise, and the following needs to be done:
- [x] `lfs` attribute with default `false`, LGTM
- [ ] figure out which files are LFS
- [ ] invoke the Git LFS filter from `$PATH`; no need for a rigid dependency or `makeWrapper`
- [ ] check the sha256
- [ ] add a test, perhaps extending `tests/nixos/fetch-git`
- [ ] documentation for the `lfs` attribute (currently under `fetchGit`'s entry in `doc/manual/src/language/builtins.md`); mention the runtime dependency on the LFS package.

kip93 commented:
What's the state on this PR? Seems to unfortunately be a bit stale given the delayed review. This issue has been plaguing us for a while, so I'm willing to pick up the torch here and try to get this out the door (was actually starting to see how to fix this myself back in March when I saw this PR and decided to see what came out of this).

b-camacho commented:
Thanks for the thorough writeup @roberth !
I owe you all an update. To avoid shelling out and implement some features we need to merge, I need a subset of a git-lft C/C++ client. I reimplemented one from Python into C++ here https://github.com/b-camacho/git-lfs-fetch-cpp.

Once I add some tests and integrate git-lfs-fetch-cpp here, we should be ready for another review!

I'm still on vacation with not-great internet, but back in 6 days and will update you all on 7/31 regardless.

Thanks for the feedback and sorry for the wait!

roberth commented:
Oh, I don't think shelling out was such a big deal because we can verify the correctness of the result, kind of like how fixed output derivations are allowed to do "grossly impure" things because we can verify the output.

I guess a library implementation of it is still nice for a consistent UX with a small closure size though.

L-as commented:
Really we need to reimplement the subset of git we use too.

L-as commented:
We however notably only want shallow fetching (_not_ `--depth`, but `--filter=tree:0`). Git doesn't do that well.

kip93 commented:
Hey! It's me again! I just want to ask if there's anything I can help with here. Maybe I can try and doing some testing, or do a smaller version of this that uses the git-lfs CLI tools while the full implementation gets done?

We have a lot of repos with LFS files that would greatly benefit from this, so I'm willing to do whatever work is needed, but also don't want to add extra work for others where it's not wanted.

b-camacho commented:
Thanks for offering to help @kip93, I'll very happily take you up on that! Here's some background on fetchers (sorry if you already know all this). `SourceAccessor` abstracts over fs-like entities. Only method we care about is 2 versions of `readFile` (one reads entire file into memory, the other streams it via callbacks). There are 2 `SourceAccessor` subclasses for git access:
![image](https://github.com/user-attachments/assets/5d4710eb-a1d9-4783-9306-c5da7b73876e)

Brief gitlog perusing does not show why `GitSourceAccessor` is not rolled into `GitExportIgnoreSourceAccessor`, but currently they are both used - so we need to make our change to both.

For now, I only added it to `GitSourceAccessor`, adding [this patch](https://gist.github.com/b-camacho/7e7b38f28b4eda1064ee8d3728d2bc67) on top of your nixpkgs then running `/where/you/built/nix/bin/nix-build -A pkgs.test-lfs` should give you an outpath with a correctly smudged file.

The major issues right now:
- `readFile` in `GitSourceAccessor`  pulls the entire file into memory. this will OOM on many many files, so we need to override the streaming version of `readFile` instead
- `GitExportIgnoreSourceAccessor` ignores LFS for now
- we do not know why `GitExportIgnoreSourceAccessor` and `GitSourceAccessor` both exist, figuring out exactly why they're both needed would be very helpful

In addition per Robert's comment above, we do still need a SHA check, some tests and docs. Tomorrow I'll investigate `GitExportIgnoreSourceAccessor` more, but I'm not very familiar with how nix handles git, so it's taking a while.

b-camacho commented:
Ok I don't think we need to touch `GitExportIgnoreSourceAccessor`! [In it's constructor](https://github.com/NixOS/nix/blob/f4962fe59b5e624dfe62a922782c6a1b44d683a4/src/libfetchers/git-utils.cc#L772), it takes an instance of `SourceAccessor` but in reality we always pass in `GitSourceAccessor` there - so the smudging will be applied either way. Aspects of GitEISA remain a mystery to me, like [which constructor is being invoked here](https://github.com/NixOS/nix/blob/f4962fe59b5e624dfe62a922782c6a1b44d683a4/src/libfetchers/git-utils.cc#L773-L775_) At first it looks like we're invoking an inherited constructor because of [this using declaration](https://github.com/NixOS/nix/blob/e1a817fb1bd8e622172800258000820d7d0c8092/src/libfetchers/filtering-source-accessor.hh#L85) but neither CFSA or FSA define any constructors that take a lambda as second arg, friends I show this to say it shouldn't compile. Yet it does.

Side quest mysteries aside, tomorrow I'll change `GitSA::readFile` to override the streaming version of `SA::readFile`.

b-camacho commented:
> neither CFSA or FSA define any constructors that take a lambda as second arg, friends I show this to say it shouldn't compile
ended up figuring this out, but it was not very interesting after all

also implemented the streaming readFile impl using `Sink` and cleaned up `git-lfs-fetch.hh`. now working on the e2e tests

b-camacho commented:
oh not super important but if you have a sec @roberth - is this the right way to run the e2e nixos tests: `nix build --impure '.#hydraJobs.tests.fetch-git' --show-trace` ? I keep seeing failures in `tests/functional/tarball.sh` but AFAIK those should not even be running for the e2e nixos-based tests

roberth commented:
> right way to run the e2e nixos tests: `nix build --impure '.#hydraJobs.tests.fetch-git' --show-trace` ?

Yes.

> I keep seeing failures in `tests/functional/tarball.sh`

The VM tests depend on the `nix` package, which runs the functional tests, or at least it [did until yesterday](https://github.com/NixOS/nix/pull/11798/files#diff-7234a357f14107533702004c573c10a11b095da0e31b309abc1f448341362f03R25), so if you rebase, it will skip `tarball.sh` as a dependency of the VM tests. (`nix-cli` is from the newly componentized packaging)

kip93 commented:
You caught me in a really busy week, but hopefully next week I should be able to work with this at my day job, so I should have the time to throw in some real work. Though it seems to me like you've advanced a lot in the last 4 days, so let me know if there's still something I can help with next week (:

b-camacho commented:
Except URL parsing quirk I think it's ready!
Added an e2e test and fixed some parsing, will be back tomorrow to touch it up.
Currently the filter attribute matching still uses my homebrew pathspec impl, I also need to replace that with https://libgit2.org/libgit2/#HEAD/group/pathspec/git_pathspec_matches_path

kip93 commented:
Seems like when you went and asked for help I could not do so and now almost everything seems to be done. It looks to me the least I can do is throw this at our repos with a variety of lfs files to see if at least I can help test that it works as expected. Looking at the nixos tests I can think of an edge case or 2 that usually git just throws warnings about, and I'm not sure how this implementation would deal with those.

b-camacho commented:
> Seems like when you went and asked for help I could not do so and now almost everything seems to be done

that's on me, I should have been more responsive when you first offered!

> least I can do is throw this at our repos with a variety of lfs files
that would be wonderful, but there is another mystery I would really like your help with @kip93! it seems _something is wrong with path caching, but only in nixos tests_.

when nix decides whether to fetch a store path or use a cached one, it uses a cache key that looks like this
```
        cacheKey = fetchers::Cache::Key{"fetchToStore", {
            {"name", std::string{name}},
            {"fingerprint", *path.accessor->fingerprint},
            {"method", std::string{method.render()}},
            {"path", path.path.abs()}
        }};
```
that `fingerprint` is what interests us - for git that's the repo rev + attributes like "useSubmodules" or "export-ignore" (and now also lfs). to get the new fingerprint, we simply [optionally append `";l"` to it](https://github.com/b-camacho/nix/blob/9a6388dce60002693b8deda1e6b0d9962f714395/src/libfetchers/git.cc#L808) when `"lfs = true"`. this part works fine on my machine and in nixos tests.


however, in nixos tests we _compute the same store path whether the `";l"` suffix is there or not_. You can verify this on your machine, pull latest of this branch and run `nix build --impure '.#hydraJobs.tests.fetch-git'` - I left in some debug prints proving the fingerprints _are_ different. conversely, if you'd like to verify that the paths are different outside of the test, you can run 
```
./outputs/out/bin/nix --extra-experimental-features nix-command eval --impure --raw --expr '(
      builtins.fetchGit {
        url = "git@gitlab.com:b-camacho/test-lfs.git";
        rev = "5536a6ce88a520ab059a04c8249cbe1f9295f7cc";
        ref = "main";
        lfs = true;
      }
    ).outPath'
./outputs/out/bin/nix --extra-experimental-features nix-command eval --impure --raw --expr '(
      builtins.fetchGit {
        url = "git@gitlab.com:b-camacho/test-lfs.git";
        rev = "5536a6ce88a520ab059a04c8249cbe1f9295f7cc";
        ref = "main";
        lfs = false;
      }
    ).outPath'
```
and observe the paths are different.

I don't know why two different fingerprints seem to produce the same cache key, especially why it only happens in the test VM, I'd love your help investigating further (I would probably start by printing the cache key?). the relevant code uses ample inheritance of implementation so it's a little hard to understand for me. Ofc if you're too busy (or would rather not look into this for any reason) lmk and I'll keep trying.

kip93 commented:
Interesting, I'll certainly have a look and let you know what I find out

kip93 commented:
Ok, just as an update of my progress (or lack thereof) so far:

* Took a while to parse through the code and try to understand the flow.
* Tried to get the cache key using `std::cerr << "nix::fetchToStore cacheKey:" << fetchers::attrsToJSON(cacheKey.value().second) << std::endl;` but it looks ok (`{"fingerprint":"8602f7662e242b18d27fdc830150e51d4fb5051b;e","method":"nar","name":"source","path":"/"}` vs `{"fingerprint":"8602f7662e242b18d27fdc830150e51d4fb5051b;e;l","method":"nar","name":"source","path":"/"}`).
* Disabled the cache code entirely, still gets the same path in both cases.
* Been going down the rabbit hole ever since, coming up with dead ends. :/
* My only suspect right now is the `treeHashToNarHash` function in `git-utils`, which assumes a NAR hash out of just a rev, but can't see where it's used and that would also fail outside of the VM? All around confusion.

Anyways, not don't have any more time this week to look at it, will try again next week

b-camacho commented:
I got hopeful and went to check `treeHashToNarHash` but it doesn't get called when running in the VM, I'm equally stumped

b-camacho commented:
Figured it out!
The storepath (ie the nix sha256) is not calculated from the fingerprint, but from contents. With `--debug` it actually says the cache missed 
```
nix eval --debug --impure --raw --expr '(
      builtins.fetchGit {
        url = "http://gitea:3000/test/lfs";
        rev = "755514e46a902f778e4ea9a087e3cafb970c1810";
        ref = "main";
        lfs = true;
      }
    ).outPath'
evaluating file '<nix/derivation-internal.nix>'
locking path '/root/.cache/nix/gitv3/0iq989hd50hql2bp8bdpn06lddbhpd2fjl21xw9qw0drrr4l5yrr'
lock acquired on '/root/.cache/nix/gitv3/0iq989hd50hql2bp8bdpn06lddbhpd2fjl21xw9qw0drrr4l5yrr.lock'
lock released on '/root/.cache/nix/gitv3/0iq989hd50hql2bp8bdpn06lddbhpd2fjl21xw9qw0drrr4l5yrr.lock'
using cache entry 'gitLastModified:{"rev":"755514e46a902f778e4ea9a087e3cafb970c1810"}' -> '{"lastModified":1731733520}'
using cache entry 'gitRevCount:{"rev":"755514e46a902f778e4ea9a087e3cafb970c1810"}' -> '{"revCount":1}'
using revision 755514e46a902f778e4ea9a087e3cafb970c1810 of repo 'http://gitea:3000/test/lfs'
smudgeLfs: 1
getFingerprint: 755514e46a902f778e4ea9a087e3cafb970c1810;e;l
did not find cache entry for 'fetchToStore:{"fingerprint":"755514e46a902f778e4ea9a087e3cafb970c1810;e;l","method":"nar","name":"source","path":"/","store":"/nix/store"}'
copying '«git+http://gitea:3000/test/lfs?exportIgnore=1&ref=main&rev=755514e46a902f778e4ea9a087e3cafb970c1810»«unknown»/' to the store...
acquiring write lock on '/nix/var/nix/temproots/1317'
/nix/store/c5a3w5jsfsrh812gdvzljgg21ccnin3a-source
```
So why was it not smudging the file? Before smudging a pointer file we check if it has a `filter=lfs` attribute in `.gitattributes`. `.gitattributes` contains lines that look like this: `*.big_ext filter=lfs...`, that first part is the "pathspec" (=~glob pattern in libgit2 lingo). When I was checking if the pathspec matches the file path, I was using an absolute, instead of relative path and the leading `/` broke the pathspec matcher.

easy fix in retrospect! Thanks so much for your help running this down @kip93!

Separately, I cleaned up the patchspec matching code and added more tests.

I think we're ready for a review, lmk what else we need @roberth!

kip93 commented:
Nice! I did not look into the smudging logic itself, so figures I did not see that u.u

Anyway, then I'll proceed to do some manual testing (I think I can also add a unit test or 2). 

Also, while waiting for compilation here and there, I wrote some small docs that are probably necessary to add. Not sure if I can make a PR into a PR, but here's the commit in my own repo: [b48dacd5](https://github.com/kip93/nix/commit/b48dacd50cbe4adf8f2849aa9144e85fbc38247a)

kip93 commented:
Just for visibility, I've already managed to break it with a not-so-much-edge-case 🙈

BUT, I'm already working on a fix, plus I already have a test in place to ensure it works as expected.

kip93 commented:
Took me a bit longer than I expected (mostly because my C++ is lacking), but I believe I've fixed the issues I could see (mostly based around people can make mistakes, don't just assume everything will just work).

Still need to test this more thoroughly with our real world usage of lfs, but from weird cases that I've seen in the wild, I think I've covered those.

b-camacho commented:
Tonight I'm gotta try `fetchGit`ing a random selection of popular lfs-using repos:
https://github.com/Wasted-Audio/hvcc
https://github.com/ChainSafe/gossamer
https://github.com/ivy-llc/ivy
https://github.com/game-ci/unity-test-runner

kip93 commented:
Again just for visibility and to avoid duplicating work, I am coming up with more errors and trying to track them down

* Sometimes we get non-zero padding errors
* I have a case where lfs=0 and lfs=1 give the same result, checking what happened here
* I am testing SSH access too and seeing issues, but this is lower prio and I believe it can be a separate PR if everything else works

kip93 commented:
Ok getting late so I should probably stop working to today :p

From my previous commits, I've fixed the issue where I found lfs=0 and lfs=1 had the same result. It turned out that in my example it was a bad repo that had lfs declared but no actual lfs files, but when I switched to another random repo I found that libgit had an issue with patterns that start with `**/`, so I filed a bug but also added a work around for now.

For the non-zero padding error, it is actually a flake specific problem. When I do fetchGit it works, but when I try to use the same repos as flake inputs it throws this error. I have added a test to ensure make sure that it works as expected.

kip93 commented:
Ok I've been trying to be transparent here so I'll admit I've not had time to look at this and won't have any tomorrow either. I can take another stab at it next week, but if you feel like trying to find the issue don't let me stop you.

b-camacho commented:
I'm happy to hop back in! Can you tell me more about the non-zero padding errors? Just a dump of your terminal output would be a lot to go off of, I don't think I ran into that one before.

kip93 commented:
[There's a test I've already added](https://github.com/NixOS/nix/blob/6778d24a27e7b1d53ad746ad1b22625faf682602/tests/nixos/fetch-git/test-cases/lfs/default.nix#L180-L200) which currently fails with this error and should explain how to reproduce, and here's the output:

```
vm-test-run-fetch-git> client # error:
vm-test-run-fetch-git> client #        … while updating the lock file of flake 'path:/build/tmpvffq_lct?lastModified=1732571537&narHash=sha256-/lQnNQ7IGCkZTsV32qCzPYSWyD1x44pl1DyaxnxR%2Blg%3D'
vm-test-run-fetch-git> client # 
vm-test-run-fetch-git> client #        … while updating the flake input 'foo'
vm-test-run-fetch-git> client # 
vm-test-run-fetch-git> client #        … while fetching the input 'git+http://gitea:3000/test/lfs?lfs=1&ref=main&rev=56afd59511b955d8e589513b58dde01cf287a364'
vm-test-run-fetch-git> client # 
vm-test-run-fetch-git> client #        error: non-zero padding
```

b-camacho commented:
Thanks for the repro! Pulled on a couple different threads, almost positive [this sizeCallback invocation](https://github.com/b-camacho/nix/blob/6778d24a27e7b1d53ad746ad1b22625faf682602/src/libfetchers/git-utils.cc#L712) is the culprit, it seems to underreport size for files that use lfs. I can confirm and fix tomorrow.

kip93 commented:
Oh that sounds very likely! If needed, I expect I'll have some more time come Wednesday.

b-camacho commented:
I was right about the size callback, the nixos test passes again!
Remerged `master`, would love to get this in before we fall too far behind again.
Last thing I can think of adding is fixes for fetching over ssh, but I agree with you it may be better to handle those in a followup.

b-camacho commented:
Hm, just re-checked I can pull from GitHub and GitLab over SSH from my machine, [this gist](https://gist.github.com/b-camacho/9aa7755917d62e40337f489b34177c0d) has the commands and outputs I ran. Current implementation intentionally uses the system `ssh` binary.

Do you have any more details on the ssh failures? You probably took this into account, but asking for completeness: even publicly available repos do not allow ssh access without auth, so you have to try this on a repo where you have owner/collaborator access on a privkey in your keychain.

Alas, while I don't exactly enjoy chasing subtle git{ea,hub,lab} impl differences, I'll try to set up an automated test for ssh fetching while we wait for a review.

kip93 commented:
Not sure what happened before (I was a bit burnt out so it might just have been a silly mistake) but now it's working fine, both public and private repos on github and gitlab work ok 🎉

Sorry for the false alarm 🙈

jsqu4re commented:
Thank you so much for your efforts! I tried this PR and it works well for `gitlab.com`. Is there a chance to also allow for local remote repos using `url = "file:///path/to/repo/.git"`? Currently there is this error:

``` bash
ssh: Could not resolve hostname file: Name or service not known
error: git-lfs-authenticate: no output (cmd: ssh file git-lfs-authenticate ///path/to/repo/.git download)
```

roberth commented on tests/nixos/fetch-git/test-cases/lfs/default.nix:
```diff
@@ -0,0 +1,202 @@
+    # memorize the revision
+    bad_lfs_rev = client.succeed(f"{repo.git} rev-parse HEAD").strip()
+
+    # prove that it can be cloned with regular git first
```
nitpick

```suggestion
    # test assumption that it can be cloned with regular git first
```

roberth commented on tests/nixos/fetch-git/test-cases/lfs/default.nix:
```diff
@@ -0,0 +1,202 @@
+    assert int(file_size_default) < expected_max_size_lfs_pointer, \
+      f"did not set lfs, yet lfs-enrolled file is {file_size_default}b (>{expected_max_size_lfs_pointer}b), probably bad default value"
+
+    # Use as flake input
+    ############################################################################
+    with TemporaryDirectory() as tempdir:
+      client.succeed(f"mkdir -p {tempdir}")
+      client.succeed(f"""
+        printf '{{
+          inputs = {{
+            foo = {{
+              url = "git+{repo.remote}?ref=main&rev={lfs_file_rev}&lfs=1";
+              flake = false;
+            }};
+          }};
+          outputs = {{ foo, self }}: {{ inherit (foo) outPath; }};
+        }}' >{tempdir}/flake.nix
+      """)
+      fetched_flake = client.succeed(f"""
+        nix eval --raw {tempdir}#.outPath
+      """)
+
+    assert fetched_lfs == fetched_flake, \
+      f"fetching as flake input (store path {fetched_flake}) yielded a different result than using fetchGit (store path {fetched_lfs})"
```
This test case seems redundant to me.

roberth commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,587 @@
+}
+
+bool Fetch::hasAttribute(const std::string & path, const std::string & attrName, const std::string & attrValue) const
+{
```
Do we have a compelling reason not to use [`git_attr_get_ext`](https://libgit2.org/docs/reference/main/attr/git_attr_get_ext.html)?

kip93 replied:
Didn't implement this part, @b-camacho may have more context, but I could try my hand at using this instead to see if it still works.

b-camacho replied:
Not a satisfying explanation but I could not get it to work without materializing the files on disk, but I'll try again

roberth reviewed (commented):
Reviewed a bit. It seems that the gitattributes support is incomplete, but could perhaps be reused from libgit2.

> Git consults [...]  the same directory as the path in question, and its parent directories up to the toplevel of the work tree

- [`man gitattributes`](https://git-scm.com/docs/gitattributes)

+ see comments

kip93 commented on tests/nixos/fetch-git/test-cases/lfs/default.nix:
```diff
@@ -0,0 +1,202 @@
+
+    # check that file was smudged
+    file_size_lfs = client.succeed(f"stat -c %s {fetched_lfs}/beeg").strip()
+    assert int(file_size_lfs) >= expected_max_size_lfs_pointer, \
```
This was a conceptual test, since a pointer cannot be >= 1024 by definition, can definitely also check for the full 1MiB size. Will rewrite this with the other test changes when I have the time

kip93 commented on src/libfetchers/git-utils.cc:
```diff
@@ -663,24 +664,82 @@ ref<GitRepo> GitRepo::openRepo(const std::filesystem::path & path, bool create,
         , root(peelToTreeOrBlob(lookupObject(*repo, hashToOID(rev)).get()))
+        , lfsFetch(lfsFetch)
     {
+        if (lfsFetch && !lookup(CanonPath(".gitattributes"))) {
+            warn("Requested to fetch lfs files, but no .gitattributes file was found, ignoring");
+        }
```
I think cloning with LFS is quite slow as it is, adding a few extra seconds at most to recurse the entire tree is probably not very noticeable

kip93 commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,587 @@
+    std::istringstream iss(content_str);
+    std::string line;
+
+    while (std::getline(iss, line)) {
```
Don't know if libgit2 has lfs support, but the syntax was simple enough that we just wrote a quick parser. I'll take another look at libgit2's provided tools next week to see if I missed anything

b-camacho replied:
git attributes exist at the git layer (as opposed to lfs), you _can_ read them with `libgit2`'s `git_attr_get` (the `ExportIgnore` fetcher does this), but I could never get `git_attr_get` to work, but I'll give it another shot tonight.

b-camacho commented:
> Thank you so much for your efforts! I tried this PR and it works well for `gitlab.com`. Is there a chance to also allow for local remote repos using `url = "file:///path/to/repo/.git"`? Currently there is this error:
> 
> ```shell
> ssh: Could not resolve hostname file: Name or service not known
> error: git-lfs-authenticate: no output (cmd: ssh file git-lfs-authenticate ///path/to/repo/.git download)
> ```

I'd be happy to add this, but I agree that his PR is already pretty big so we should save it for a followup.

jsqu4re commented:
> I'd be happy to add this, but I agree this PR is already pretty big so we should save it for a followup.

Makes total sense, and thanks for tackling this task!

CHN-beta commented:
> > Thank you so much for your efforts! I tried this PR and it works well for `gitlab.com`. Is there a chance to also allow for local remote repos using `url = "file:///path/to/repo/.git"`? Currently there is this error:
> > ```shell
> > ssh: Could not resolve hostname file: Name or service not known
> > error: git-lfs-authenticate: no output (cmd: ssh file git-lfs-authenticate ///path/to/repo/.git download)
> > ```
> 
> I'd be happy to add this, but I agree this PR is already pretty big so we should save it for a followup.

Seems git LFS is only supported over https now. Enabling other transfers (e.g. ssh) needing https://github.com/NixOS/nixpkgs/pull/350130. Please let me know if I am wrong.

kip93 commented:
Today I'll try to rewrite all tests, but wanted to let you know that the guys from libgit2 actually replied to my [issue](https://github.com/libgit2/libgit2/issues/6946), and in there I've found out that this function we're using (`git_pathspec_matches_path`) is not the right one (and there's currently no right function for this job) since a pathspec is not the same as a path pattern. So I'll try to use the `git_attr_get_ext` again and see if that works for our purposes. This should also fix issues with negated patterns that I've noticed today.

b-camacho commented:
Oh I had no idea they were different, that explains a lot!

kip93 commented:
Ok I've reintroduced the git_attr_get_ext and have it working. I'm currently testing that it does not import files into the store too early like it used to be the case the first time around.

Also rewrote the tests to use subtests as suggested by the code review.

kip93 commented:
With my quick-ish tests (with several large lfs repos which take a long time to clone) it all looks ok? I think it's ready for another review.

Given that holidays are around the corner, I won't have any more time to work on this 'till after new years (of course, I'm also not expecting you to skip xmas to review this, I'm fine if the review is delayed as well). If anything needs fixing I can take this back up again probably 6-ish of january.

edolstra commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,440 @@
+
+// if authHeader is "", downloadToSink assumes no auth is expected
+void downloadToSink(
+    const std::string & url, const std::string & authHeader, Sink & sink, std::string_view sha256Expected)
```
This should use `FileTransfer::download()` (the sink variant). That way libfetcher doesn't need to reimplement curl support, and we get stuff like connection reuse.

edolstra commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,440 @@
+    size_t size;      // in bytes
+};
+
+struct GitUrl
```
This class looks very similar to `ParsedUrl`, maybe you can use that instead (in conjuction with `fixGitURL()` to handle the scp syntax)?

edolstra commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,440 @@
+    return jArray;
+}
+
+std::vector<nlohmann::json> Fetch::fetchUrls(const std::vector<Md> & metadatas) const
```
Idem, this should use `FileTransfer::download()`.

kip93 replied:
Since this is an API call, I implemented this with FileTransfer::upload and added a "post" flag like there was a "head" flag to make it work

b-camacho commented on src/libfetchers/git-utils.cc:
```diff
@@ -669,24 +670,72 @@ ref<GitRepo> GitRepo::openRepo(const std::filesystem::path & path, bool create,
+        if (lfsFetch) {
+            auto& _lfsFetch = *lfsFetch;
+            auto pathStr = std::string(path.rel());
+            if (_lfsFetch.shouldFetch(pathStr)) {
+                StringSink s;
+                try {
+                    _lfsFetch.fetch(blob.get(), pathStr, s, [&s](uint64_t size){ s.s.reserve(size); });
```
I was mimicking `if let Some(v) = v {}`, but in C++ pointer-like syntax is more idiomatic for optionals

kip93 commented:
Hey! Been back at it for a couple of days now, trying to implement the code reviews. Currently, only the reworks of Fetch::fetchUrls is left, but I might or might not be down a rabbit hole since the FileTransfer::download (nor the FileTransfer::upload) quite do what I need it to. Will let you guys know when it's all done.

kip93 commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,440 @@
+        }
+        const uint64_t size = obj.at("size");
+        sizeCallback(size);
+        downloadToSink(ourl, authHeader, sink, oid); // oid is also the sha256
```
Since this does not have the store to be able to put things there and it is much more closely related to git I ended up putting it in the user's cache dir instead

kip93 commented:
Been testing it, looks like it's working for me. I'd say this is ready for another review.

roberth commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,329 @@
+    } catch (const nlohmann::json::out_of_range & e) {
+        std::stringstream ss;
+        ss << "bad json from /info/lfs/objects/batch: " << obj << " " << e.what();
+        throw std::runtime_error(ss.str());
```
Please change all of these to `Error`

edolstra commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,329 @@
+    } catch (const nlohmann::json::parse_error & e) {
+        std::stringstream ss;
+        ss << "response did not parse as json: " << responseString;
+        throw std::runtime_error(ss.str());
```
In fact with `Error` you probably don't need the `std::stringstream`, i.e.
```
throw Error("response did not parse as json: %s", responseString);
```

kip93 commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,329 @@
+struct Fetch
+{
+    // Reference to the repository
+    git_repository const * repo;
```
Tried this, but could not get it to work. Something about sizeof incomplete type. Not a c++ dev and I could not figure it out but for what I understand since we're linking to an external library here the compiler might not have enough info to compile this?

kip93 commented on src/libfetchers/git-lfs-fetch.hh:
```diff
@@ -0,0 +1,329 @@
+    auto responseString = result.data;
+
+    std::vector<nlohmann::json> objects;
+    // example resp here:
+    // {"objects":[{"oid":"f5e02aa71e67f41d79023a128ca35bad86cf7b6656967bfe0884b3a3c4325eaf","size":10000000,"actions":{"download":{"href":"https://gitlab.com/b-camacho/test-lfs.git/gitlab-lfs/objects/f5e02aa71e67f41d79023a128ca35bad86cf7b6656967bfe0884b3a3c4325eaf","header":{"Authorization":"Basic
+    // Yi1jYW1hY2hvOmV5SjBlWEFpT2lKS1YxUWlMQ0poYkdjaU9pSklVekkxTmlKOS5leUprWVhSaElqcDdJbUZqZEc5eUlqb2lZaTFqWVcxaFkyaHZJbjBzSW1wMGFTSTZJbUptTURZNFpXVTFMVEprWmpVdE5HWm1ZUzFpWWpRMExUSXpNVEV3WVRReU1qWmtaaUlzSW1saGRDSTZNVGN4TkRZeE16ZzBOU3dpYm1KbUlqb3hOekUwTmpFek9EUXdMQ0psZUhBaU9qRTNNVFEyTWpFd05EVjkuZk9yMDNkYjBWSTFXQzFZaTBKRmJUNnJTTHJPZlBwVW9lYllkT0NQZlJ4QQ=="}}},"authenticated":true}]}
```
To test this function would require at least an http server and for full test also an ssh server, tried to find if there's some other test I can reference from but none that I could find.

edolstra reviewed (approved):
Did some code review/cleanup, otherwise looks great! Thanks!

roberth commented:
Thank you!

It would be great to also have this in `inputs.self`, like
- https://github.com/NixOS/nix/pull/12421

kip93 commented:
I can make a PR for that next week (:

kip93 commented:
Well I lied I found a little bit of time to do this today and it turned out to be much simpler than I thought https://github.com/NixOS/nix/pull/12468

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-2-27-0-released/62003/1
